math/cosine.go

```diff
@@ -1,73 +1,65 @@
-SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
-SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
-HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
-OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
-SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-*/
 import (
 	stdMath "math"
+	"math/bits"
 
 	"github.com/ericlagergren/decimal"
+	"github.com/ericlagergren/decimal/internal/arith/checked"
+	"github.com/ericlagergren/decimal/misc"
 )
 
-var cosine4NMaxN = uint64((stdMath.Sqrt(4.0*float64(stdMath.MaxUint64)+1.0) + 1.0) / 4.0)
+const maxInt = 1<<(bits.UintSize-1) - 1 // Also: uint64(int(^uint(0) << 1))
 
-func prepareCosineInput(precision int, xValue *decimal.Big) (*decimal.Big, int, bool) {
-	c2Pi := Pi(decimal.WithPrecision(precision))
-	c2Pi = c2Pi.Mul(c2Pi, two)
-	var x *decimal.Big
-	calculatingPrecision := precision
-	//we need to make sure the value we're working with a value is closer to zero (for better results)
-	if c2Pi.CmpAbs(xValue) < 0 {
-		//for cos to work correctly the input
-		// must be |xValue|< 2Pi so we'll fix it
+func prepCosine(z, x *decimal.Big, ctx decimal.Context) (*decimal.Big, int, bool) {
+	x0 := alias(z, x)
 
-		v := decimal.WithPrecision(calculatingPrecision).QuoInt(xValue, c2Pi)
-		vInt, ok := v.Int64()
+	var tmp decimal.Big
+
+	// for better results, we need to make sure the value we're working with a
+	// value is closer to zero.
+	ctx.Mul(&tmp, Pi(&tmp, ctx), two) // 2 * Pi
+	if x.CmpAbs(&tmp) >= 0 {
+		// for cos to work correctly the input must be in (-2Pi, 2Pi).
+		ctx.Quo(&tmp, x, &tmp)
+		misc.SetSignbit(&tmp, false)
+		v, ok := tmp.Uint64()
 		if !ok {
 			return nil, 0, false
 		}
-		//now we'll resize Pi to be a more accurate precision
-		piPrecision := precision + int(stdMath.Ceil(stdMath.Abs(float64(vInt))/float64(10)))
-		piMultiplier := decimal.New(2*vInt, 0)
-		toRemove := Pi(decimal.WithPrecision(piPrecision))
-		toRemove = toRemove.Mul(toRemove, piMultiplier)
-		//so toRemove = 2*Pi*vInt so
-		// xValue-toRemove < 2*Pi
-		//add 1 to the precision for the up comming squaring
-		x = decimal.WithPrecision(calculatingPrecision).Sub(xValue, toRemove)
+
+		// Adjust so we have ceil(v/10) + ctx.Precision, but check for overflows.
+		// 1+((v-1)/10) will be widly incorrect for v == 0, but x/y = 0 iff
+		// x = 0 and y != 0. In this case, -2pi <= x >= 2pi, so we're fine.
+		prec, ok := checked.Add(1+((v-1)/10), uint64(ctx.Precision))
+		if !ok || prec > maxInt {
+			return nil, 0, false
+		}
+		pctx := decimal.Context{Precision: int(prec)}
+
+		if v <= stdMath.MaxUint64/2 {
+			tmp.SetUint64(2 * v)
+		} else {
+			ctx.Mul(&tmp, &tmp, two)
+		}
+
+		var pi decimal.Big
+		pctx.Mul(&tmp, Pi(&pi, pctx), &tmp)
```

Comment:
tmp lost the proper sign on line 25

Reply:
yeah, it's the wrong precision though. I didn't like it either, but we need the initial Pi to get the extra precision which we then use to generate _another_ instance of Pi. Got any ideas?

Reply:
Turns out there is also a bug here. But the failure is mine. I tested in my code manually in my other project, but never wrote a test for it in the decimal.Big project. 
We can get rid of the bug and deal with the 2pi thing at that same time:

```
	var tmp decimal.Big
	var tmp2pi decimal.Big

	// for better results, we need to make sure the value we're working with a
	// value is closer to zero.
	ctx.Mul(&tmp2pi, Pi(&tmp2pi, ctx), two) // 2 * Pi
	if x.CmpAbs(&tmp2pi) >= 0 {
		// for cos to work correctly the input must be in (-2Pi, 2Pi).
		ctx.QuoInt(&tmp, x, &tmp2pi)
		v, ok := tmp.Uint64()
		if !ok {
			return nil, 0, false
		}

		// Adjust so we have ceil(v/10) + ctx.Precision, but check for overflows.
		// 1+((v-1)/10) will be wildly incorrect for v == 0, but x/y = 0 iff
		// x = 0 and y != 0. In this case, -2pi <= x >= 2pi, so we're fine.
		prec, ok := checked.Add(1+((v-1)/10), uint64(ctx.Precision))
		if !ok || prec > maxInt {
			return nil, 0, false
		}
		pctx := decimal.Context{Precision: int(prec)}

		if v <= stdMath.MaxUint64/2 {
			tmp.SetUint64(v)
		} 

		pctx.Mul(&tmp, &tmp2pi, &tmp)

		// so toRemove = 2*Pi*v so x - toRemove < 2*Pi
		ctx.Sub(x0, x, &tmp)
	}
```

Reply:
Ah, I was wondering why that instance differed from the other one in a different function that did a similar thing.

Reply:
I just corrected the above code.  to calculate prec it needed to be pos but the .Uint64() return zero i think when the value is negative.

Reply:
Sounds good. Question: why'd you remove the `ctx.Mul(&tmp, &tmp, two)` line (for when `v > stdMath.MaxUint64/2`)?

Reply:
The `two` on  line 42 `ctx.Mul(&tmp, &tmp, two)` was only needed because the later on line 47 `pctx.Mul(&tmp, Pi(&pi, pctx), &tmp)` you were multiplying by Pi when it needed to be 2pi.  But by using a variable to store 2pi it's not needed. Note line 41 `tmp.SetUint64(2 * v)` was also changed by removing the unneeded 2.